apps/prairielearn/python/colors.py

```diff
@@ -1,196 +1,139 @@
+
+
+class PrarieLearnColor(sRGB):
+    """Custom sRGB class to handle custom PrarieLearn colors, via Coloraide."""
+
+    def match(self, string, start=0, fullmatch=True):
+        """
+        Match a color string, first trying PrarieLearn.
+        If no match is found, defaults to sRGB class' implementation.
+        """
+        # Match the string using fullmatch if requested
+        match = (
+            RE_PL_COLORS.fullmatch(string, start)
+            if fullmatch
+            else RE_PL_COLORS.match(string, start)
+        )
+
+        if match:
+            # See if we can find the name
+            name = match.group(1)
+            values = PL_COLORS_NAME_MAP.get(name.lower(), None)
+
+            if values is not None:
+                # Normalize back to 0 - 1
+                values = [c / 255 for c in values]
+                # Return the coordinates and transparency separate,
+                # also include the match end position.
+                return (values[:-1], values[-1]), match.end(1)
+
+        # Couldn't find custom colors, so use the default color matching.
+        return super().match(string, start, fullmatch)
+
+    def to_string(
+        self,
+        parent,
+        *,
+        alpha=None,
+        precision=None,
+        fit=True,
+        none=False,
+        color=False,
+        hex=False,  # pylint: disable=redefined-builtin
+        names=False,
+        comma=False,
+        upper=False,
+        percent=False,
+        compress=False,
+        **kwargs,
+    ) -> str:
+        """Convert to string."""
+
+        if names:
+            # Get alpha and coordinates resolving undefined values as required
+            alpha = serialize.get_alpha(parent, alpha, False, False)
+            if alpha is None:
+                alpha = 1
+            coords = serialize.get_coords(parent, fit, False, False) + [alpha]
+
+            # See if the color value is a match, if so, return the string
+            result = PL_COLORS_VALUE_MAP.get(
+                tuple([alg.round_half_up(c * 255) for c in coords]), None
+            )
+            if result is not None:
+                return result
+
+        # Color name not requested, or there was no color match, use default serialization.
+        return super().to_string(
+            parent,
+            alpha=alpha,
+            precision=precision,
+            fit=fit,
+            none=none,
+            color=color,
+            hex=hex,
+            names=names,
+            comma=comma,
+            upper=upper,
+            percent=percent,
+            compress=compress,
+            **kwargs,
+        )
 
-def get_css_color(name: str) -> Optional[str]:
-    """
-    Tries to look up a hex code value from a named css color, otherwise will
-    return None if not a valid color.
-    """
 
-    name = name.lower()
-    if name in colors:
-        return colors[name]
-    else:
-        return None
+Color.register(PrarieLearnColor(), overwrite=True)
```

Comment:
Minor issue (I think): this code will get called every time this file is imported. The standard way of resolving this is just by wrapping all of this in a function (maybe something like `get_color_class`). I'm not sure if there's a way to get around this that doesn't substantially change the developer semantics though, so maybe not an issue (also because this register call probably isn't very expensive).

Reply:
I still haven't address this, but for what it's worth, a very similar `.register` call happens in the actual `coloraide` source ([link](https://github.com/facelessuser/coloraide/blob/6f5874e934b18d0cb558ad15a52acbce2f2a9cb5/coloraide/color.py#L1279)). Perhaps it's okay?

Reply:
Apparently, my comment subscribed me here. Once you import, even if you import it again, it is already imported. It is registered only once. The only way this would get run multiple times is if you created multiple instances of Python that then import, then it would be registered in each instance. A single instance of Python will import once and be done.